composable/sequential.go

```diff
@@ -0,0 +1,204 @@
+package composable
+
+import (
+	"context"
+	"errors"
+	"sync"
+
+	"github.com/ipfs/go-cid"
+	"github.com/libp2p/go-libp2p-core/peer"
+	"github.com/libp2p/go-libp2p-core/routing"
+	"github.com/multiformats/go-multiaddr"
+)
+
+var _ routing.Routing = &Sequential{}
+
+type Sequential struct {
+	routers []*SequentialRouter
+}
+
+func NewSequential(routers []*SequentialRouter) *Sequential {
+	return &Sequential{
+		routers: routers,
+	}
+}
+
+func (r *Sequential) Provide(ctx context.Context, cid cid.Cid, provide bool) error {
+	for _, router := range r.routers {
+		ctx, cancel := context.WithTimeout(ctx, router.Timeout)
+		if err := router.Router.Provide(ctx, cid, provide); err != nil &&
+			!errors.Is(err, routing.ErrNotSupported) &&
+			!router.IgnoreError {
+			cancel()
+			return err
+		}
+		cancel()
+	}
+
+	return nil
+}
+
+func (r *Sequential) FindProvidersAsync(ctx context.Context, cid cid.Cid, count int) <-chan peer.AddrInfo {
+	chanOut := make(chan peer.AddrInfo)
+	sentCount := 0
+	go func() {
+		for _, router := range r.routers {
+			ctx, cancel := context.WithTimeout(ctx, router.Timeout)
+			rch := router.Router.FindProvidersAsync(ctx, cid, count)
+
+			var g sync.WaitGroup
+			g.Add(1)
+			go func() {
+				defer g.Done()
+				for {
+					select {
+					case <-ctx.Done():
+						return
+					case v, ok := <-rch:
+						if !ok {
+							break
+						}
+						if sentCount >= count {
+							break
+						}
+
+						chanOut <- v
+						sentCount++
+					}
+				}
+			}()
+
+			g.Wait()
+			cancel()
+		}
+	}()
+
+	return chanOut
+}
+
+func (r *Sequential) FindPeer(ctx context.Context, pid peer.ID) (peer.AddrInfo, error) {
+	maSet := make(map[multiaddr.Multiaddr]struct{})
```

Comment:
@aschmahmann @guseggert Does putting all different multi addresses together makes sense?

Reply:
IIUC folks would use sequential when they want to short-circuit after a result is found, which I think you changed this code to right?

Reply:
fixed